src/infra/dataverse/sparql/sparqlGateway.ts

```diff
@@ -0,0 +1,142 @@
+    limit: number,
+    offset: number,
+    filters: RetrieveDataverseQueryFilters
+  ): TE.TaskEither<Error, RetrieveDataverseResult> => {
+    const query = `
+      PREFIX core: <https://ontology.okp4.space/core/>
+      PREFIX owl: <http://www.w3.org/2002/07/owl#>
+      PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
+      PREFIX serviceMetadata: <https://ontology.okp4.space/metadata/service/>
+      PREFIX datasetMetadata: <https://ontology.okp4.space/metadata/dataset/>
+      PREFIX dataspaceMetadata: <https://ontology.okp4.space/metadata/dataspace/>
+      SELECT ?id ?metadata ?title ?description ?type
+      WHERE {
+        {
+          ?id rdf:type core:Service .
+          ?metadata rdf:type serviceMetadata:GeneralMetadata .
+        } UNION {
+          ?id rdf:type core:Dataset .
+          ?metadata rdf:type datasetMetadata:GeneralMetadata .
+        } UNION {          
+          ?id rdf:type core:DataSpace .
+          ?metadata rdf:type dataspaceMetadata:GeneralMetadata .
+        }
+        ${filters.byType !== 'all' ? `FILTER ( contains(str(?type), "${filters.byType}" ))` : ''}
+        ?id rdf:type ?type .
+        ?type rdf:type owl:Class .
+        ?metadata core:describes ?id .
+        ?metadata core:hasTitle ?title .
+        ?metadata core:hasDescription ?description .
+        FILTER ( langMatches(lang(?title),'${language}') && langMatches(lang(?description),'${language}') )
+      }
+      LIMIT ${limit}
+      OFFSET ${offset}
+`
+
+    const fetchHeaders: RequestInit = {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/x-www-form-urlencoded',
+        Authorization: `Basic ${APP_ENV.sparql['credentials']}`,
+        accept: 'application/sparql-results+json'
+      },
+      body: `query=${encodeURIComponent(query)}`
+    }
+
+    const fetchDataverse = (): TE.TaskEither<Error, Response> =>
+      TE.tryCatch(
+        async () => fetch(APP_ENV.sparql['endpoint'], fetchHeaders),
+        reason =>
+          reason instanceof Error
+            ? reason
+            : new Error(`Oops.. Something went wrong fetching ontology: ${JSON.stringify(reason)}`)
+      )
+
+    const serializeResponse = (response: Response): TE.TaskEither<Error, SparqlResult> =>
+      TE.tryCatch(
+        async () => response.json(),
+        reason =>
+          reason instanceof Error
+            ? reason
+            : new Error(
+                `Oops.. Something went wrong serializing sparql response: ${JSON.stringify(reason)}`
+              )
+      )
+
+    const splitBindingId = (b: SparqlBinding): O.Option<[SparqlBinding, string]> =>
+      pipe(
+        O.fromNullable(b.id.value.split('/').at(-1)),
+        O.map(a => [b, a])
+      )
+
+    const splitBindingType = ([b, a]: [SparqlBinding, string]): O.Option<
+      [SparqlBinding, string, string]
+    > =>
+      pipe(
+        O.fromNullable(b.type.value.split('/').at(-1)),
+        O.map(c => [b, a, c])
+      )
```

Comment:
Can be refactored a bit:

```ts
    // this can be put in utils
    const getURILastElement = (str: string): O.Option<string> =>
      O.fromNullable(str.split('/').at(-1))

    const splitBindingId = (b: SparqlBinding): O.Option<[SparqlBinding, string]> =>
      pipe(
        b.id.value,
        getURILastElement,
        O.map(a => [b, a])
      )

    const splitBindingType = ([b, a]: [SparqlBinding, string]): O.Option<
      [SparqlBinding, string, string]
    > =>
      pipe(
        b.type.value,
        getURILastElement,
        O.map(c => [b, a, c])
      )
```

Reply:
Yep, much more clearer.